src/coreclr/vm/prestub.cpp

```diff
@@ -1032,6 +1042,629 @@ PCODE MethodDesc::JitCompileCodeLocked(PrepareCodeConfig* pConfig, COR_ILMETHOD_
     return pCode;
 }
 
+bool MethodDesc::TryGenerateTransientILImplementation(DynamicResolver** resolver, COR_ILMETHOD_DECODER** methodILDecoder)
+{
+    if (TryGenerateAsyncThunk(resolver, methodILDecoder))
+    {
+        return true;
+    }
+
+    if (TryGenerateUnsafeAccessor(resolver, methodILDecoder))
+    {
+        return true;
+    }
+
+    return false;
+}
+
+bool MethodDesc::TryGenerateAsyncThunk(DynamicResolver** resolver, COR_ILMETHOD_DECODER** methodILDecoder)
+{
+    STANDARD_VM_CONTRACT;
+    _ASSERTE(resolver != NULL);
+    _ASSERTE(methodILDecoder != NULL);
+    _ASSERTE(*resolver == NULL && *methodILDecoder == NULL);
+    _ASSERTE(IsIL());
+    _ASSERTE(GetRVA() == 0);
+
+    if (!IsAsyncThunkMethod())
+    {
+        return false;
+    }
+
+    MethodDesc *pAsyncOtherVariant = this->GetAsyncOtherVariant();
```

Comment:
Here we create a Task-returning variant for an Async method. It is more code than the other way around, but essentially it calls the Async method and then examines if we have a continuation returned or an actual result and calls a helper that wraps that into an appropriate `Task`.